File: packages/bdd/src/diagnostics.ts
import type { Span } from './span.js'

export type Severity = 'error' | 'warning'

export type Diagnostic = {
  readonly severity: Severity
  readonly code: DiagnosticCode
  readonly message: string
  readonly span: Span
}

export type DiagnosticCode = 'ambiguous-match'

export type Candidate = {
  readonly expression: string
  readonly sourceFile: string
  readonly sourceLine: number
}

export type AmbiguousInput = {
  readonly text: string
  readonly span: Span
  readonly candidates: ReadonlyArray<Candidate>
}

export function ambiguousMatch(input: AmbiguousInput): Diagnostic {
  const lines = input.candidates
    .map((c) => `  step('${c.expression}', ...)    at ${c.sourceFile}:${c.sourceLine}`)
    .join('\n')
  return {
    severity: 'error',
    code: 'ambiguous-match',
    message: `Ambiguous step: "${input.text}"\nMatched by:\n${lines}`,
    span: input.span,
  }
}

File: python/packages/var-core/src/var_core/diagnostics.py
"""diagnostics.py — port of var-core/src/diagnostics.ts.

Only the subset needed by the planner is ported here:
  Severity, DiagnosticCode, Diagnostic, Candidate, AmbiguousInput,
  ambiguous_match(), error_fence_without_step().
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from var_core.span import Span

Severity = Literal["error", "warning"]
DiagnosticCode = Literal["ambiguous-match", "error-fence-without-step"]


@dataclass(frozen=True, slots=True)
class Candidate:
    expression: str
    source_file: str
    source_line: int


@dataclass(frozen=True, slots=True)
class Diagnostic:
    code: DiagnosticCode
    severity: Severity
    message: str
    span: Span


@dataclass(frozen=True, slots=True)
class AmbiguousInput:
    text: str
    span: Span
    candidates: tuple[Candidate, ...]


def ambiguous_match(input: AmbiguousInput) -> Diagnostic:
    """Mirror ambiguousMatch() from diagnostics.ts."""
    lines = "\n".join(
        f"  '{c.expression}'    at {c.source_file}:{c.source_line}"
        for c in input.candidates
    )
    return Diagnostic(
        severity="error",
        code="ambiguous-match",
        message=f'Ambiguous step: "{input.text}"\nMatched by:\n{lines}',
        span=input.span,
    )


def error_fence_without_step(span: Span) -> Diagnostic:
    """Mirror errorFenceWithoutStep() from diagnostics.ts."""
    return Diagnostic(
        severity="error",
        code="error-fence-without-step",
        message=(
            "This `error` fence marks the example as expected-to-fail, "
            "but the example has no step to run."
        ),
        span=span,
    )
